src/Test/QuickCheck/Gen.hs

```diff
@@ -293,6 +294,16 @@ gen `suchThatMaybe` p = sized (\n -> try n (2*n))
         x <- resize m gen
         if p x then return (Just x) else try (m+1) n
 
+-- | Tries to generate a value that satisfies a predicate.
+-- If it fails to do so it discards the test case if the result
+-- is used in the test.
+suchThatDiscard :: Gen a -> (a -> Bool) -> Gen a
+suchThatDiscard g p = do
+  a <- g
```

Comment:
Another option here is `(\ a -> if p a then a else discard) <$> g` - which I suppose is _slightly_ more lazy than this alternative?